[quant][graphmode] `insert_quant_dequant` pass support shared class types

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types**
* #31407 [quant][graphmode][refactor] Remove observers in the end
* #31406 [quant][graphmode][refactor] Call `getQSchemeAndQParamMap` later in `quantizeTensors`
* #31293 [quant][graphmode][refactor] Get QScheme from observer module
* #31292 [refactor] Remove redundant queries of qconfig in `insertObservers`

Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

jerryzh168 committed:
[quant][graphmode] `insert_quant_dequant` pass support shared class types

Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1242,6 +1242,48 @@ def forward(self, x, w0, w1, w2):
+
+            get_forward(m._c)(data)
+            m._c = torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False)
+            # assert len(m._modules._c.items()) == 1, \
```
Looks like a leftover, please remove.

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -1242,6 +1242,48 @@ def forward(self, x, w0, w1, w2):
+            get_forward(m._c)(data)
+            quant_func = "aten::quantize_per_channel" if is_per_channel \
+                else "aten::quantize_per_tensor"
+            FileCheck().check_not(quant_func) \
```
Please add a comment of what exactly are we checking in this test. I.e. to me it looks that the current check is verifying that we're inserting quantize calls inside the inner method. It is not clear, however, how it verifies that everything works correctly for shared types.

jerryzh168 replied:
Yeah, this PR is still WIP :)

ZolotukhinM replied:
Please re-request review when it's ready!

ZolotukhinM reviewed (approved):
Looks good, but please include the test into the PR where the functional change was introduced.

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 commented:
Oh this is ready, please take a look

Sent from my iPhone

On Dec 8, 2019, at 13:48, Mikhail Zolotukhin <notifications@github.com> wrote:

﻿

@ZolotukhinM commented on this pull request.

________________________________

In test/test_jit.py<https://github.com/pytorch/pytorch/pull/30553#discussion_r355217306>:

> +            qconfig = QConfig(activation=observer, weight=observer)
+            qconfig_dict = {
+                '': script_qconfig(qconfig)
+            }
+            m._c = torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, False)
+            data = torch.randn(1, 3, 10, 10, dtype=torch.float)
+
+            get_forward(m._c)(data)
+            m._c = torch._C._jit_pass_insert_quant_dequant(m._c, "forward", False)
+            # assert len(m._modules._c.items()) == 1, \
+            #    'Expected to have single submodule of conv'
+
+            get_forward(m._c)(data)
+            quant_func = "aten::quantize_per_channel" if is_per_channel \
+                else "aten::quantize_per_tensor"
+            FileCheck().check_not(quant_func) \


Please re-request review when it's ready!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/30553?email_source=notifications&email_token=ABF2R2KM43MAFSCMJU5DLNTQXVTQNA5CNFSM4JSN6662YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOLJCKQ#discussion_r355217306>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABF2R2M4KHL777YNUPSW5ZLQXVTQNANCNFSM4JSN666Q>.

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -835,6 +835,21 @@ void InsertQuantDeQuantHelper::run(
+      for (const auto& pr : qparams) {
+        const auto& name = pr.first;
+        const auto& qparam = pr.second;
+        module._ivalue()->setAttr(v->debugName() + name, qparam);
```
Have you checked that it works for a module with several submodules of the same type? If m has conv1 and conv2, their qparams can be different, but this code will always set the same values to them for each instance of the conv module. I think there must be some calls to `get_qparams` somewhere here.

jerryzh168 replied:
Yeah I'm aware that it doesn't work, I plan to add the support in a separate PR

jerryzh168 replied:
In that case, we need to have different graphs for each type of qparams, since different qparams might correspond to different quantize calls

ZolotukhinM replied:
Why would two convs have separate graphs? The graphs would be the same, and only the data (attributes) of the particular modules would differ. Could you please elaborate?

jerryzh168 replied:
I'm talking about quantized graph, input graph is the same for conv module, but if we have want to quantize one conv module instance using per tensor affine and the other using per channel affine, then we'll need to produce two different quantized graphs.

ZolotukhinM replied:
What if I just want to quantize all convs the same way and I have 2+ of them. I was talking about this case, which seemed to me more frequent - in this case the graph would be the same and the current code will not give correct results.

jerryzh168 replied:
Oh, I see, yeah, `values_to_qparams` should be indexed by Module as well, I'll change this

jerryzh168 commented:
@ZolotukhinM I want to enable simple cases first and then support more complicated case.

ZolotukhinM commented:
> @ZolotukhinM I want to enable simple cases first and then support more complicated case.

I see, I don't see any rush in landing this until the more complicated case is supported though - until then this would be unusable in real scenarios anyway (i.e. most probably it will yield a wrong result when we have two conv submodules). Incomplete solutions should not be landed unless we can at least error out on non-supported cases.

jerryzh168 commented:
> > @ZolotukhinM I want to enable simple cases first and then support more complicated case.
> 
> I see, I don't see any rush in landing this until the more complicated case is supported though - until then this would be unusable in real scenarios anyway (i.e. most probably it will yield a wrong result when we have two conv submodules). Incomplete solutions should not be landed unless we can at least error out on non-supported cases.
I think there is a misunderstanding here, it doesn't support the case when we have two conv submodules that shares the same number of input/output channels and the qconfig for them generate different graphs, e.g. one per tensor quant, one per channel quant. I think this is a PR to cover most of the cases already.

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
support modules with shared class type, but with same qconfig, we'll support the case with different qconfig in future PR.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 83a3529:
* **2/17** broken upstream at merge base 20a2e52 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=20a2e526efb353e302f5b8778b63e45a5eade240&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * Since your merge base is older than `viable/strict`, run these commands:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **15/17** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=83a35299acdccbde5898900270248526ac18dfa2).

### 8 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956023) pytorch_linux_xenial_py3_clang5_mobile_build (1/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479216))
<details>
<summary>
<code>Dec 15 01:46:02 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return</code>
</summary>

```
Dec 15 01:46:02                  from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:46:02 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:46:02  enum class QScheme : uint8_t { 
Dec 15 01:46:02             ^ 
Dec 15 01:46:02 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:46:02 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:46:02 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:46:02  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:46:02                               ^ 
Dec 15 01:46:02 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp: In member function 'int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)': 
ror: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return 
Dec 15 01:46:02    return std::make_tuple(qscheme, qparams); 
Dec 15 01:46:02                                           ^ 
Dec 15 01:46:02 caffe2/CMakeFiles/torch_cpu.dir/build.make:8536: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:46:02 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:46:02 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:46:02 [ 98%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/fuser/fallback.cpp.o 
Dec 15 01:46:02 CMakeFiles/Makefile2:882: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:46:02 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:46:02 Makefile:138: recipe for target 'all' failed 
Dec 15 01:46:02 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956012) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479448))
<details>
<summary>
<code>Dec 15 01:48:11 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return</code>
</summary>

```
Dec 15 01:48:11                  from /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:48:11 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:11  enum class QScheme : uint8_t { 
Dec 15 01:48:11             ^ 
Dec 15 01:48:11 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:11 /var/lib/jenkins/workspace/c10/core/QScheme.h:15:12: note:   'c10::QScheme' 
Dec 15 01:48:11 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:48:11  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:48:11                               ^ 
Dec 15 01:48:11 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp: In member function 'int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)': 
ror: cannot convert 'std::tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >' to 'int' in return 
Dec 15 01:48:11    return std::make_tuple(qscheme, qparams); 
Dec 15 01:48:11                                           ^ 
Dec 15 01:48:11 caffe2/CMakeFiles/torch_cpu.dir/build.make:20050: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:48:11 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:48:11 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:48:11 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:48:11 CMakeFiles/Makefile2:5422: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:48:11 Makefile:138: recipe for target 'all' failed 
Dec 15 01:48:11 make: *** [all] Error 2 
Dec 15 01:48:11 Traceback (most recent call last): 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955896) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (3/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479464))
<details>
<summary>
<code>Dec 15 01:48:15 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:10: error: no viable conversion from returned value of type 'tuple<typename __decay_and_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int'</code>
</summary>

```
Dec 15 01:48:15   if (tuple->elements().size() != num_outputs) { 
Dec 15 01:48:15       ~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~ 
Dec 15 01:48:15 1 warning generated. 
Dec 15 01:48:15 [ 78%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/mobile/import.cpp.o 
Dec 15 01:48:15 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 15 01:48:15   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 15 01:48:15              ^ 
Dec 15 01:48:15 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 15 01:48:15 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:48:15            ^ 
nd_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int' 
Dec 15 01:48:15   return std::make_tuple(qscheme, qparams); 
Dec 15 01:48:15          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Dec 15 01:48:15 3 errors generated. 
Dec 15 01:48:15 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:48:15 caffe2/CMakeFiles/torch_cpu.dir/build.make:20050: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:48:15 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:48:15 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:48:15 CMakeFiles/Makefile2:5548: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:48:15 Makefile:138: recipe for target 'all' failed 
Dec 15 01:48:15 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956019) pytorch_linux_xenial_py3_clang5_asan_build (4/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74479542))
<details>
<summary>
<code>Dec 15 01:48:33 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:771:10: error: no viable conversion from returned value of type 'tuple<typename __decay_and_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int'</code>
</summary>

```
Dec 15 01:48:32 [ 83%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/mobile/module.cpp.o 
Dec 15 01:48:32 [ 83%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/mobile/register_mobile_ops.cpp.o 
Dec 15 01:48:33 [ 83%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/mobile/interpreter.cpp.o 
Dec 15 01:48:33 [ 83%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/autograd/VariableTypeManual.cpp.o 
Dec 15 01:48:33 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 15 01:48:33   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 15 01:48:33              ^ 
Dec 15 01:48:33 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 15 01:48:33 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:48:33            ^ 
nd_strip<QScheme &>::__type, typename __decay_and_strip<unordered_map<basic_string<char>, IValue, hash<string>, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::__type>' (aka 'tuple<c10::QScheme, std::unordered_map<std::__cxx11::basic_string<char>, c10::IValue, std::hash<std::__cxx11::string>, std::equal_to<std::__cxx11::basic_string<char> >, std::allocator<std::pair<const std::__cxx11::basic_string<char>, c10::IValue> > > >') to function return type 'int' 
Dec 15 01:48:33   return std::make_tuple(qscheme, qparams); 
Dec 15 01:48:33          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Dec 15 01:48:33 3 errors generated. 
Dec 15 01:48:33 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:48:33 make[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:48:33 caffe2/CMakeFiles/torch_cpu.dir/build.make:20770: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o' failed 
Dec 15 01:48:33 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 15 01:48:33 CMakeFiles/Makefile2:5242: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 15 01:48:33 Makefile:138: recipe for target 'all' failed 
Dec 15 01:48:33 make: *** [all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956013) pytorch_macos_10_13_cuda10_0_cudnn7_py3_build (5/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74480861))
<details>
<summary>
<code>Dec 14 17:53:09 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Dec 14 17:53:06 [2503/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/canonicalize_modified_loop.cpp.o 
Dec 14 17:53:07 [2504/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/edit_distance.cpp.o 
Dec 14 17:53:07 [2505/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/logging.cpp.o 
Dec 14 17:53:07 [2506/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/module.cpp.o 
Dec 14 17:53:08 [2507/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/object.cpp.o 
Dec 14 17:53:08 [2508/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/jit_exception.cpp.o 
Dec 14 17:53:08 [2509/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/source_range_serialization.cpp.o 
Dec 14 17:53:08 [2510/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/tracer.cpp.o 
Dec 14 17:53:09 [2511/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/hooks_for_testing.cpp.o 
Dec 14 17:53:09 [2512/2966] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Dec 14 17:53:09 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
READ -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c ../torch/csrc/jit/passes/quantization.cpp 
Dec 14 17:53:09 ../torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 14 17:53:09   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 14 17:53:09              ^ 
Dec 14 17:53:09 ../torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 14 17:53:09 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 14 17:53:09            ^ 
me __make_tuple_return<unordered_map<basic_string<char>, IValue, hash<basic_string<char> >, equal_to<basic_string<char> >, allocator<pair<const basic_string<char>, IValue> > > &>::type>' (aka 'tuple<c10::QScheme, std::__1::unordered_map<std::__1::basic_string<char>, c10::IValue, std::__1::hash<std::__1::basic_string<char> >, std::__1::equal_to<std::__1::basic_string<char> >, std::__1::allocator<std::__1::pair<const std::__1::basic_string<char>, c10::IValue> > > >') to function return type 'int' 
Dec 14 17:53:09   return std::make_tuple(qscheme, qparams); 
Dec 14 17:53:09          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956003) pytorch_windows_build (6/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74481051))
<details>
<summary>
<code>FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.obj</code>
</summary>

```

HAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\fuser\codegen.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\fuser\codegen.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

-DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\utils\byte_order.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\utils\byte_order.cpp 
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\passes\quantization.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\passes\quantization.cpp 
FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.obj  
_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -O2 -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -std:c++14 /showIncludes /Focaffe2\CMakeFiles\torch_cpu.dir\__\torch\csrc\jit\passes\quantization.cpp.obj /Fdcaffe2\CMakeFiles\torch_cpu.dir\ /FS -c ..\torch\csrc\jit\passes\quantization.cpp 
..\torch\csrc\jit\passes\quantization.cpp(624): error C2065: 'QScheme': undeclared identifier
..\torch\csrc\jit\passes\quantization.cpp(624): error C3544: '_Types': parameter pack expects a type template argument
..\torch\csrc\jit\passes\quantization.cpp(661): error C2027: use of undefined type 'std::tuple'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\tuple(197): note: see declaration of 'std::tuple'
..\torch\csrc\jit\passes\quantization.cpp(662): error C2672: 'std::get': no matching overloaded function found
..\torch\csrc\jit\passes\quantization.cpp(662): error C2784: 'tuple_element<_Idx,std::pair<_Ty1,_Ty2>>::type &std::get(std::pair<_Ty1,_Ty2> &) noexcept': could not deduce template argument for 'std::pair<_Ty1,_Ty2> &' from 'std::tuple'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.23.28105\include\utility(453): note: see declaration of 'std::get'
..\torch\csrc\jit\passes\quantization.cpp(663): error C3536: 'qparam_map': cannot be used before it is initialized
ue>>>,std::_Uhash_compare<_Kty,_Hasher,_Keyeq>,_Alloc,false>>::insert(std::pair<const _Kty,std::unordered_map<std::string,c10::IValue,std::hash<std::basic_string<char,std::char_traits<char>,std::allocator<char>>>,std::equal_to<std::basic_string<char,std::char_traits<char>,std::allocator<char>>>,std::allocator<std::pair<const std::basic_string<char,std::char_traits<char>,std::allocator<char>>,c10::IValue>>>> &&)': cannot convert argument 1 from 'initializer list' to 'std::pair<const _Kty,_Ty> &&'
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956022) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (7/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74481542))
<details>
<summary>
<code>Dec 15 01:54:35 /pytorch/torch/csrc/jit/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return</code>
</summary>

```
Dec 15 01:54:35                  from /pytorch/torch/csrc/jit/passes/quantization.cpp:1: 
Dec 15 01:54:35 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:35  enum class QScheme : uint8_t { 
Dec 15 01:54:35             ^~~~~~~ 
Dec 15 01:54:35 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:35 /pytorch/c10/core/QScheme.h:15:12: note:   ‘c10::QScheme’ 
Dec 15 01:54:35 /pytorch/torch/csrc/jit/passes/quantization.cpp:740:30: error: template argument 1 is invalid 
Dec 15 01:54:35  std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 01:54:35                               ^ 
Dec 15 01:54:35 /pytorch/torch/csrc/jit/passes/quantization.cpp: In member function ‘int torch::jit::{anonymous}::InsertQuantDeQuantHelper::getQSchemeAndQParamMap(torch::jit::script::Module&, torch::jit::Value*)’: 
t/passes/quantization.cpp:771:42: error: cannot convert ‘std::tuple<c10::QScheme, std::unordered_map<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue> > > >’ to ‘int’ in return 
Dec 15 01:54:35    return std::make_tuple(qscheme, qparams); 
Dec 15 01:54:35                                           ^ 
Dec 15 01:54:37 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/compiler.cpp.o 
Dec 15 01:54:37 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/testing/file_check.cpp.o 
Dec 15 01:54:38 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/convert_to_ssa.cpp.o 
Dec 15 01:54:38 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/exit_transforms.cpp.o 
Dec 15 01:54:39 [ 80%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/script/inline_loop_condition.cpp.o 
Dec 15 01:54:39 gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o] Error 1 
Dec 15 01:54:39 gmake[2]: *** Waiting for unfinished jobs.... 
Dec 15 01:54:58 gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3955995) pytorch_linux_xenial_py3_clang5_android_ndk_r19c_x86_32_build (8/8)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74484723))
<details>
<summary>
<code>Dec 15 02:05:37 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o</code>
</summary>

```
Dec 15 02:05:08 [533/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/remove_inplace_ops.cpp.o 
Dec 15 02:05:14 [534/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/specialize_autogradzero.cpp.o 
Dec 15 02:05:19 [535/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/subgraph_rewrite.cpp.o 
Dec 15 02:05:27 [536/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/shape_analysis.cpp.o 
Dec 15 02:05:27 [537/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/python_print.cpp.o 
Dec 15 02:05:28 [538/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/subgraph_utils.cpp.o 
Dec 15 02:05:29 [539/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/utils/memory_dag.cpp.o 
Dec 15 02:05:34 [540/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/fuse_linear.cpp.o 
Dec 15 02:05:35 [541/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/print_handler.cpp.o 
Dec 15 02:05:37 [542/581] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o 
Dec 15 02:05:37 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o  
rameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/jit/passes/quantization.cpp.o -c /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp 
Dec 15 02:05:37 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:624:14: error: use of undeclared identifier 'QScheme' 
Dec 15 02:05:37   std::tuple<QScheme, QParamMap> getQSchemeAndQParamMap(script::Module& module, Value* v); 
Dec 15 02:05:37              ^ 
Dec 15 02:05:37 /var/lib/jenkins/workspace/torch/csrc/jit/passes/quantization.cpp:740:12: error: use of undeclared identifier 'QScheme' 
Dec 15 02:05:37 std::tuple<QScheme, QParamMap> InsertQuantDeQuantHelper::getQSchemeAndQParamMap( 
Dec 15 02:05:37            ^ 
ndk1::char_traits<char>, std::__ndk1::allocator<char> >, c10::IValue, std::__ndk1::hash<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::equal_to<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > >, std::__ndk1::allocator<std::__ndk1::pair<const std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, c10::IValue> > > >') to function return type 'int' 
Dec 15 02:05:37   return std::make_tuple(qscheme, qparams); 
Dec 15 02:05:37          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
```

</details>

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build](https://dr.pytorch.org/build-details.html?build_id=74482743)

### 8 failures *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py2_7_9_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_6_gcc5_4_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_xla_linux_xenial_py3_6_clang7_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_macos_10_13_py3_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_cuda9_cudnn7_py3_build</sup> | <sup>Build</sup> | &#128721; [Broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=20a2e526efb353e302f5b8778b63e45a5eade240&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35&highlight_job=pytorch_linux_xenial_cuda9_cudnn7_py3_build) |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>binary_linux_manywheel_2_7mu_cpu_devtoolset7_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build</sup> | <sup>Build</sup> | New in PR |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>binary_macos_libtorch_2_7_cpu_build</sup> | <sup>Build</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
support modules with shared class type, but with same qconfig, we'll support the case with different qconfig in future PR.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
att, we'll error out when a graph is quantized with different QSchemes
This is a rare case, so it should be OK to skip for now. Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 commented:
@ZolotukhinM just updated, please take a look again

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -661,8 +667,15 @@ void InsertQuantDeQuantHelper::collectObserverNodesAndValueToQuantize(
+  }
   auto qparam_map = std::get<1>(tp);
-  values_to_qparams_[g].insert({new_value, qparam_map});
+  values_to_qparams_[g][module._ivalue()].insert({new_value, qparam_map});
```
This indexing by a module object seems weird. Can we just set the proper attribute in `module._ivalue()` and then use it? The map would then be to the name of the attribute I think.

jerryzh168 replied:
I realized that we shouldn't store qparam here, we should call `getQSchemeAndQParamMap` for each module instance instead

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] `insert_quant_dequant` pass support shared class types"


Summary:
We'll error out when a graph is quantized with different QSchemes.
This only occurs when we have two modules that have same types (e.g. two Conv2d modules initialized with
same arguments) and quantized with two configs that would produce different quantized graphs, for example
per tensor affine and per channel affine. This is a rare case, so it should be OK to skip for now.
Actual support will come later.

Test Plan:
test_jit.py, test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18849022](https://our.internmc.facebook.com/intern/diff/D18849022)

[ghstack-poisoned]

jerryzh168 commented:
this is closed somehow, I'll create again

jerryzh168 commented:
see https://github.com/pytorch/pytorch/pull/31408
